File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

   public ListNode reverse(ListNode head){
    ListNode tempp = head;
    ListNode prev = null;

    while(tempp!= null){
        ListNode nxt = tempp.next;
        tempp.next= prev;
        prev=tempp;
        tempp= nxt;
    }
    return prev;
   }

    public boolean isPalindrome(ListNode head) {
        ListNode temp = head;
        ListNode slow = head;
        ListNode fast = head;

        while(fast!= null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;
        }
        ListNode newHead= reverse(slow);

        ListNode i = head;
        ListNode j = newHead;
        

        while(j!=null){
        if(i.val!= j.val){
            return false;
        }
        i= i.next;
        j= j.next;
    }
       return true;
    }
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def reverse(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev  = None
        curr = head
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        return prev    

    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return True
        
        slow = head 
        fast = head
        while fast and fast.next:
            slow = slow.next 
            fast = fast.next.next
            
        j =self.reverse(slow)
        i = head
        
        while j:
           if i.val != j.val:
             return False  
           i = i.next
           j = j.next  
        return True 